### Exercise 1.1

##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [13]:
import pandas as pd
from tables.leaf import limit_es
import requests

response = requests.get("https://data.cityofnewyork.us/resource/2upf-qytp.json",
                        params={'$limit':50_000})

if response.ok:
  payload=response.json()
else:
  print(f"Request was not successful and returned code: {response.status_code}.")
  payload = None

df = pd.DataFrame(payload)
df.head(5)

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime passenger_count  \
0        2  2019-08-19T15:04:10.000  2019-08-20T14:26:20.000               1   
1        1  2019-08-19T21:39:58.000  2019-08-19T22:19:53.000               2   
2        2  2019-08-19T21:10:34.000  2019-08-19T21:13:05.000               1   
3        2  2019-08-19T21:14:20.000  2019-08-19T21:20:11.000               1   
4        2  2019-08-19T21:42:54.000  2019-08-20T21:20:11.000               1   

  trip_distance ratecodeid store_and_fwd_flag pulocationid dolocationid  \
0          5.22          1                  N          231          181   
1         18.10          2                  N          132           48   
2          0.58          1                  N           79          107   
3          1.31          1                  N          107          113   
4          7.17          1                  N          148           24   

  payment_type fare_amount extra mta_tax tip_amount tolls_amount  \
0            1          19   0.5     0.5       4.56            0   
1            2          52   2.5     0.5          0         6.12   
2            1           4   0.5     0.5       1.56            0   
3            1         6.5   0.5     0.5       2.06            0   
4            2        23.5   0.5     0.5          0            0   

  improvement_surcharge total_amount congestion_surcharge  
0                   0.3        27.36                  2.5  
1                   0.3        61.42                  2.5  
2                   0.3         9.36                  2.5  
3                   0.3        12.36                  2.5  
4                   0.3         27.3                  2.5

### Exercise 1.2

##### Find the dimensions (number of rows and number of columns) in the data.

In [14]:
df.to_csv('my_taxis.csv')

df.shape

(50000, 18)

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

---

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.